Fix Maven plugin help task

aSemy opened the pull request:
Fix #3035

- update `pom.template.xml` so the HelpMojo is generated correctly
- Created basic test for `dokka:help`
- split up the tasks that prepare/generate/sync the Maven plugin files
- remove (now) unused `MavenCliSetupExtension.mavenBuildDir`

aSemy commented:
hey @binkley, could you take a look please?

binkley commented:
Do I have the right Dokka version?
I updated to 1.8.20 after dependabot noticed.
However the release page says this is a beta version:
https://github.com/Kotlin/dokka/releases/tag/v1.8.20

aSemy commented:
> Do I have the right Dokka version? I updated to 1.8.20 after dependabot noticed. However the release page says this is a beta version: https://github.com/Kotlin/dokka/releases/tag/v1.8.20

Hey, sorry, I'm not sure what you mean?

This PR is a proposed fix, it's not merged into the master branch yet. Even after it's merged, it could be a few months before it's released!

But I would appreciate it if you reviewed my proposed changes, and ask questions or suggest improvements. You could also check out my branch, publish it to Maven Local `./gradlew publishToMavenLocal`, and then try using 1.9.0-SNAPSHOT in your project. I don't work with Maven any more, so I'm only guessing at how to fix it.

aSemy commented on runners/maven-plugin/api/maven-plugin.api:
```diff
@@ -94,6 +94,11 @@ public final class org/jetbrains/dokka/maven/ExternalDocumentationLinkBuilder {
 	public final fun setUrl (Ljava/net/URL;)V
 }
 
+public class org/jetbrains/dokka/maven/HelpMojo : org/apache/maven/plugin/AbstractMojo {
+	public fun <init> ()V
+	public fun execute ()V
+}
+
```
This addition indicates that the help task hasn't worked at most since Dokka 1.4.32 ([when the BCV was introduced](https://github.com/Kotlin/dokka/pull/1774)), and probably didn't before that.

binkley commented:
Apologies that this conversation is going _off topic_. The **real purpose** is that `mvn dokka:help` does the right thing

@aSemy Sorry, I do not understand.
I expect the release pages to talk about public versions, and betas and release candidates (something I value).
What I do not expect is that a beta release is discussed as a final release. 
It violates the _principle of least astonishment_ (https://en.wikipedia.org/wiki/Principle_of_least_astonishment)

So if a git tag in the repo refers to a release version (ie, something like x.y.z) when pointing to beta, this is a problem. The release notes refer to the tag of 1.8.20 but the release notes state this is Beta.

IgnatBeresnev commented:
> I expect the release pages to talk about public versions, and betas and release candidates (something I value).
> What I do not expect is that a beta release is discussed as a final release.

Dokka as a tool is technically in Beta now, see [Stability of Kotlin components](https://kotlinlang.org/docs/components-stability.html). Not sure what you mean by "the final release", but I agree it can be confusing. Hopefully it won't be a problem for much longer though, as Dokka is slowly approaching the Stable release.

___

I don't know much about Maven plugins, but I can have a look at the PR if you think it's ready @aSemy. I think some of the bugs you mentioned might be due to the outdated versions of Maven dependencies, I don't remember when we updated them last

binkley commented:
@IgnatBeresnev Thank you for the explanation. My expectations were in the wrong place. I think of Dokka as a solid tool, and forgot about the beta nature.

aSemy commented:
> I don't know much about Maven plugins, but I can have a look at the PR if you think it's ready @aSemy. I think some of the bugs you mentioned might be due to the outdated versions of Maven dependencies, I don't remember when we updated them last

Please take a look! I thought I would attempt a fix since I refactored the Maven Plugin config in https://github.com/Kotlin/dokka/pull/2911.

IgnatBeresnev reviewed (approved):
Great job, as always :+1:

IgnatBeresnev commented:
Could you please rebase onto master (or merge it in) to resolve the conflicts? After that, I'll run the integration tests and make sure publishing is not affected, and it can be merged

aSemy commented on runners/maven-plugin/build.gradle.kts:
```diff
@@ -43,77 +42,113 @@ val generatePom by tasks.registering(Sync::class) {
+                """.trimIndent()
+            } else {
+                line
             }
```
Fixed by randomly copy-pasting things from StackOverflow into `pom.template.xml` until the `@Mojo` annotation is generated.

IgnatBeresnev commented:
Ran the test locally, everything looks good. Thank you very much for the contribution!
